### DISCUSSION TITLE
실습코드 4-11 WebSecurityConfig 에러 문의

### DISCUSSION THREAD

**Discussion by jwkcp** [fsoftwareengineer/todo-application#39] [질문과 답변] [answered]
안녕하세요.

### 환경
SpringBoot 2.6.0
Intellij 2021.03

### 문제
[실습코드 4-11] http.addFilterAfter 부분에서 아래와 같은 에러가 발생합니다. 
**http.addFilterAfter** 부분을 삭제하면 에러가 발생하지 않습니다. 

```
@Override
    protected void configure(HttpSecurity http) throws Exception {
        // http 시큐리티 빌더
        http.cors() // WebMvcConfig에서 이미 설정했으므로 기본 cors 설정.
                .and()
                .csrf()// csrf는 현재 사용하지 않으므로 disable
                .disable()
                .httpBasic()// token을 사용하므로 basic 인증 disable
                .disable()
                .sessionManagement()  // session 기반이 아님을 선언
                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
                .and()
                .authorizeRequests() // /와 /auth/** 경로는 인증 안해도 됨.
                .antMatchers("/", "/auth/**").permitAll()
                .anyRequest() // /와 /auth/**이외의 모든 경로는 인증 해야됨.
                .authenticated();

        // filter 등록.
        // 매 리퀘스트마다
        // CorsFilter 실행한 후에
        // jwtAuthenticationFilter 실행한다.
        http.addFilterAfter(
                jwtAuthenticationFilter,
                CorsFilter.class
        );
    }
```

### 질문
addFilterAfter를 구현하는 부분이 집필 당시와 뭔가 바뀐 부분이 있는지 궁금합니다.

### 상세
오류가 나는 곳을 따라가보면 아래 [this.filterOrders.getOrder(registeredFilter)]가 null을 리턴합니다.

```
private HttpSecurity addFilterAtOffsetOf(Filter filter, int offset, Class<? extends Filter> registeredFilter) {
    int order = this.filterOrders.getOrder(registeredFilter) + offset;
    this.filters.add(new OrderedFilter(filter, order));
    this.filterOrders.put(filter.getClass(), order);
    return this;
}

```

### 오류 메시지
Error creating bean with name 'springSecurityFilterChain' defined in class path resource [org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.servlet.Filter]: Factory method 'springSecurityFilterChain' threw exception; nested exception is java.lang.IllegalArgumentException: Cannot register after unregistered Filter class org.apache.catalina.filters.CorsFilter

**Comment by southbell2** [comment, marked as answer]
저도 비슷한 에러가 났었는데 CorsFilter 의 import 가 제대로 됐는지 확인해보세요 이게 import 하는게 2개가 있어서 책이랑 다른게 import 됐을수도 있어요

**Reply by jwkcp** [reply]
아, 너무 감사한 답변입니다. 고맙습니다.
말씀대로 아래와 같이 잘못된 CorsFilter가 Import 되어 있었습니다.

[**BAD**] import org.apache.catalina.filters.CorsFilter;
[**GOOD**] import org.springframework.web.filter.CorsFilter;

감사합니다.

**Reply by mthekid** [reply]
답변 감사합니다. 엄청 해매고 있었는데... import 처리가 잘못된거였네요.

**Reply by lutae2000** [reply]
저도 해메고 있었는데 답변 감사합니다

**Reply by jwan-kim1589** [reply]
우왕 ~~~생각도 못한 부분에서 해결되었네요 감사합니다.

**Reply by fenomeno** [reply]
와 3시간만에 찾았네...

**Reply by skyriv213** [reply]
와 감사합니다....덕분에 바로 해결했어요

**Comment by fsoftwareengineer** [comment]
위의 방법으로 해결되지 않는경우 다른 필터 클래스의 다음에 추가하셔도 됩니다. 

예) CorsFilter.class 대신 UsernamePasswordAuthenticationFilter.class 사용

http.addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class);

**Reply by jwkcp** [reply]
책 감사히 잘 보고 있습니다. 이런 좋은 공간 마련해주셔서 감사합니다.
